### DISCUSSION TITLE
inaccurate SOC coming from Daly BMS

### DISCUSSION THREAD

**Discussion by hansgeorgk** [Louisvdw/dbus-serialbattery#152] [Q&A]
Hi,

I thought I have asked that question last week already - but cannot find it any more ...
I recently emptied my batteries because of an inaccurate SOC coming from Daly BMS. Daly BMS cannot measure current lower than 1 or 2 AMPS. So after a few weeks, while I always thinking I have a fully Battery at 99% it was in reality getting more and empty and finally went off 
![Bildschirmfoto 2022-07-15 um 14 52 26](https://user-images.githubusercontent.com/100286224/179226755-f1094f68-9c31-4515-9136-8568a89cd2ae.png)

The reason for this had been the small currents over weeks, which are not measured/recognized by Daly BMS

@Louisvdw is there a change to use the SOC of the BMV in your software?

In VRM I can choose which SOW should be shown in the Dashboard
Settings -> System Setup -> Batterie monitor
![Bildschirmfoto 2022-07-15 um 14 56 21](https://user-images.githubusercontent.com/100286224/179227380-007bb63d-9335-48c4-9287-de14f7709ae4.png)

Would be cool to have similar selection option in SerialBattery Drive

thanks
cheers
HG

**Comment by hansgeorgk** [comment]
![Bildschirmfoto 2022-07-15 um 17 59 12](https://user-images.githubusercontent.com/100286224/179267703-af1ba97c-97e0-4552-a992-25db6871cc3e.png)
 forgot to add the SOC screenshot. As you can see SOC is 70%, when the system Voltage is dropping and the Battery finally Is off. I had to manually set the SOW to a lower value, but it was a guess from the amount of AmpH loaded into the Battery ...

**Comment by Louisvdw** [comment]
Hi Hans
Unfortunately Daly BMS are not the best at this. I prefer any of the other BMS as they manage and balance cells better.

The driver does not connect to the BMV, only to the BMS. So it does not know about that SOC, but by setting the BMV as the battery monitor it's SOC value will be used.

**Comment by hansgeorgk** [comment]
so the setting shown above should solve the problem - right?
Or are there multiple places to set the battery monitor?

**Reply by Louisvdw** [reply]
Yes it will help for now. By setting the BMV as the battery monitor it's SOC will be used. But the BMV does not know what is happening inside your battery. It will only solve your drift problem for your GX. Even the BMV will drift a bit, but it is much better calibrated out of the factory.
Your BMS will still think the SOC is something else and your cells will still do the same as before, as your BMS is managing your cells.

The better option would be to calibrate your BMS using the BMV. (I don't know the procedure for the Daly - sorry)
Many people add external balancers to the Daly, and disable the internal balancer.
Try to frequently use your battery on large current charge/discharge. Most BMS will have reset values on spesific voltage intervals to correct your SOC values.

**Reply by SanderV4n** [reply]
Hi Louis,

I have the same issues, the soc of the BMS (Daly Sinowealth) is alway close to 100%, it does not register small loads. My BMV seems to be better at registering this and shows a better soc.

So I tried setting the the BMV as battery monitor, but the charge current stays at 5A (as the soc of the bms is at 99%). So for me this did not work. Did I miss something? I restarted the multiplus, but this also did not have an effect.

**Comment by hansgeorgk** [comment]
Hi Louis,
I like your CCCM feature. But with the incorrect SOC coming from the Daly BMS, CCCM is limiting charge current (above 91%) or even stops charging if above 99%
_# Change depending on the SOC values
        if self.soc > 99:
            self.control_allow_charge = False_
even if the battery might be far away from 91 or 99%

If I try to switch on SVS - nothing happens. SCS is de-activated. 
In STS I can choose the temperature from the different devices connected. 
![Bildschirmfoto 2022-07-26 um 13 39 06](https://user-images.githubusercontent.com/100286224/180997426-0ede1c08-df6c-41a9-b199-33ebc165fe66.png)
It might be helpful, if I could choose SVSand/or SCS ... ?
Any other idea to get a workaround for the inaccurate Daly BMS SOC?

thanks
cheers
HG

**Reply by geromd** [reply]
@Louisvdw Hi, my Daly BMS is running now after managing some problems with the hardware and USB-Setup by galvanic isolation and now I am coming to the same issue wit different SOC meassurment of Daly-BMS and Victron Multiplus integrated SOC meassurement. Daly is giving 99% SOC while Victron internal calculation is giving 94% to the battery monitor. The current limitation to 5A is starting based on the BMS SOC and much earlier than it would work out with the SOC from Victron. Could it be a solution to change the calculation of the charge current limitation optional to a trigger of an adjustable max. cell voltage instead of the unsteady BMS SOC calculation?

**Comment by hansgeorgk** [comment]
Hi Louis,

I looked at your code in battery.py:
```
  _def manage_charge_current(self):
        # Start with the current values

        # Change depending on the SOC values
```
As mentioned earlier, I very much like your way of managing the charge current the closer it gets to a full battery
Wouldn't it be possible to manage charge current based on the voltage instead of using the SOC?
eg, stop charging at 3.45V, limit charge current from 3.3
< 3.3 100%
3.3-3.35 50%
3.35-3.4 25%
3.4-3.45 5A
charging false above 3.45

What do you think?
cheers
HG

**Reply by ganey** [reply]
Can you grab the correct voltage from the BMS/config? as the voltage is different between a lipo and lifepo4 bms

**Reply by Louisvdw** [reply]
Hi @hansgeorgk 
When I originally created the CCCM I did just that and looked at the cell voltages only, but it did not work well. 
Actually I was reducing the charge voltage instead of limiting the current, but it had the same issue.
The problem is that the voltage for lithium cells can jump very fast. Too fast for the system to react. 

I am thinking that doing some soft of hybrid check might work, where if the cell voltages are all below 3.35V and the SOC is 100% that the reduction is not that high. 
If the SOC is more acurate though, that would also solve this. It could be worth while playing with the BMS settings for capacity to see how that influence it.

**Reply by Louisvdw** [reply]
@ganey that is also very valid considiration as not all the cell types use the same voltages.

**Comment by geromd** [comment]
@hansgeorgk Bevor I used the driver my DALY BMS was cutting the Charge-Connection sometimes, due that the voltage of some cells was running into the voltage limit, when loading with high current at the end. A side effect was, that the DALY-BMS set the SOC automatically to 100% when reached such a limit. Maybe this can be a trigger for a work arround?

**Reply by Louisvdw** [reply]
That is exactly how the CCCM will help when you reach 100% SOC. 
The problem seem to be that some BMS are not as good as other in their SOC calculation which then sometimes limit healthy balanced cells. If we could improve the settings of the batteries so that the SOC is more accurate then this request is not needed anymore. 

Most BMS will jump to 100% SOC when a cell imbalance happens and one cell jumps high.

**Comment by hansgeorgk** [comment]
@Louisvdw - have you looked into the hybrid check you suggested sow weeks ago?
Could we eg change the following line in battery.py 
original:
        if self.soc > 99:                    
            self.control_allow_charge = False

suggested additional check for the voltage
	if self.soc > 99 and self.cell.voltage >= MAX_CELL_VOLTAGE:                    
            self.control_allow_charge = False

or suggested check for the voltage only
	if self.cell.voltage >= MAX_CELL_VOLTAGE:                    
            self.control_allow_charge = False

so CCCM would only turn charging off, once the max voltage is reached. 
Not sure if I am using the right variables in my suggested code change ...

**Reply by Louisvdw** [reply]
Not yet.
